nebu/bakedpdf.py

```diff
@@ -0,0 +1,129 @@
+from lxml import etree
+
+from cnxepub.models import Binder, TranslucentBinder, Document
+from cnxepub.formatters import SingleHTMLFormatter
+
+
+__all__ = (
+    'assembler',
+)
+
+
+"""`Document`s require that modules contain a body tag, but this assembler
+allows for building a single/collection HTML with empty modules, therefore
+we must use this variable in place of data when building `Document` models.
+When updating this also update the tests.
+"""
+DATA_WHEN_MODULE_NOT_FOUND = "<body>Module with id {} goes here.</body>"
+
+
+def assembler(source_dir, out_dir='.', out_fname='collection'):
```

Comment:
The nature of this is a little odd. You've written this as a function that utilizes a bunch of internal functions. This is going to make tracebacks hard to follow. Also, unit testing this is difficult because you essentially need a full usecase for all corner cases.

Here is my suggestion:

```
# `assembler` becomes...
class Assembler(object):
     # where you have defined `main` make it `__call__`
     def __call__(self, source_dir, out_dir='.', out_fname='collection'):
        ...
     # def <other-functions-as-methods>(self, ...):

assembler = Assembler()  # Singleton
```

You're then free to create the `Adapter` class outside of this class. Then the `Adapter` class can be tested independent of this function.

Reply:
Note, this change won't adjust the behavior of your logic. It will simply make the individual components easier to read and reason able.

Reply:
I understand. However, IMO, the way I've written it out has at least a couple advantages:

1. Encapsulation - it clearly shows that there are 0 external dependencies on anything other than the `assembler` function. I find that capturing/following the dependencies train in CNX is a **MAJOR** pain so I have no intention of adding to the problem.
2. Separation Of Concerns - what I would've done if I wanted to not deviate from the style of coding in Neb is put everything in a single function—essentially—and just add comments above/before each chunk of code which seems to kinda be doing one thing only with the data coming from above in order to create a **visual** distinction in an _attempt_ to make code easier to read and reason about.

Besides, as far as I'm aware there no unit tests in all of the code (at least each one I've looked at) in CNX besides the ones I created for my `class PressElement` anyway, they're all integration tests or system tests, TBH.

In short, I've considered your suggestion but the way I've coded it is intentional and it's all implementation details which can later on at any point be refactored _easily_ if you so wish 😉 and per our recent meeting with @edwoodward, I must say that the right thing to do is to create a separate issue to refactor my code here—after merging it.

Reply:
In addition, I do agree that my code is imperfect or odd, even, (no pun intended) but if/when it grows that would be a great opportunity to refactor if necessary.